Introduce `Prefix` validated type

passaro opened the pull request:
Follow-up to #162. Introduces a new `Prefix` type which contains a valid string (either empty or ending in `/`) to use as a prefix argument in `S3Filesystem`, `S3FuseFilesystem`, and `Superblock`. Uses the new type for the `--prefix` argument and returns a meaningful error if not valid.

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and I agree to the terms of the [Developer Certificate of Origin (DCO)](https://developercertificate.org/).

passaro commented:
I'm looking for feedback on a couple of points:
* I added `Prefix` to `inode.rs` since it's used by `Superblock`. Is it the right place?
* I've converted the argument from `Option<String>` to `Option<Prefix>` and propagated the latter downstream. This makes it clear that the prefix is optional at every level, but may be redundant: `None` and `Some(Prefix { "" })` are equivalent. Would it be preferable to have the empty string to be the default value for `Prefix` (and remove `Option<>`)?

eslrahc-swa commented on mountpoint-s3/src/inode.rs:
```diff
@@ -839,6 +837,57 @@ pub enum InodeError {
     InodeNotReadableWhileWriting(InodeNo),
 }
 
+mod prefix {
```
Not sure if we should define prefix in a separate file and assign the related path in `lib.rs` ?

passaro replied:
Moved to prefix.rs - but happy to change if there is a better location.

passaro commented:
>     * I've converted the argument from `Option<String>` to `Option<Prefix>` and propagated the latter downstream. This makes it clear that the prefix is optional at every level, but may be redundant: `None` and `Some(Prefix { "" })` are equivalent. Would it be preferable to have the empty string to be the default value for `Prefix` (and remove `Option<>`)?

I've now switched from Option<Prefix> to Prefix (and implemented Default).

dannycjones reviewed (commented):
Nice! I like this change, as we know now that a `Prefix` is always safe to use without any validation checks.

Only minor comments.

dannycjones commented on mountpoint-s3/src/inode.rs:
```diff
@@ -839,6 +837,50 @@ pub enum InodeError {
+            }
+        }
+
+        pub fn is_valid(prefix: &str) -> bool {
+            prefix.is_empty() || prefix.ends_with('/')
+        }
```
We should add a few test cases for this too.

- `ashlbafshasfbhjla`
- ``
- `/`
- `//`
- `/hello/`
- `hello/`

Then this validation behavior is documented through tests.

passaro replied:
About the `is_valid` method: my intent was to make it a public method to check whether a string is a valid prefix (and keep the logic in one place). One (the only?) example is in Superblock::lookup, which I have now modified.
Thoughts?

jamesbornholt commented on mountpoint-s3/src/prefix.rs:
```diff
@@ -0,0 +1,65 @@
+
+use anyhow::anyhow;
+
+/// A prefix string ending in `/`, or the empty string
+#[derive(Debug, Clone, Default)]
+pub struct Prefix {
+    path: String,
+}
```
I think you're right that we should just treat `None` and `Prefix("")` the same way, so we should get rid of `Option<Prefix>` everywhere. That also simplifies a few other things.

passaro replied:
Removed last usage of Option<Prefix> in CliArgs (using default_value = "" instead)

jamesbornholt commented on mountpoint-s3/src/inode.rs:
```diff
@@ -127,7 +126,7 @@ impl Superblock {
             return Err(InodeError::NotADirectory(parent_ino));
         }
         let mut full_path = parent.full_key().to_owned();
-        assert!(full_path.is_empty() || full_path.ends_with('/'));
+        assert!(Prefix::is_valid(&full_path));
```
I wouldn't change this—to me they feel similar only coincidentally; a `Prefix` is "a subset of the bucket" whereas this is more like "a query I'm about to send to S3". I can imagine worlds where we might change the behavior of `Prefix` in the future (for example, to no longer require ending in a `/`).

jamesbornholt commented on mountpoint-s3/src/prefix.rs:
```diff
@@ -0,0 +1,65 @@
+use std::fmt::Display;
+use std::str::FromStr;
+
+use anyhow::anyhow;
```
I'd build a real error type here instead of using anyhow:
```rust
use thiserror::Error;

#[derive(Error, Debug)]
pub enum PrefixError {
    #[error("prefixes must end in '/'")]
    MissingFinalDelimiter,
}
```
and then
```rust
impl Prefix {
    pub fn new(prefix: &str) -> Result<Self, PrefixError> {
        ....
```

(Roughly speaking, our approach has been "real error types everywhere except `main.rs`", which means anyone using this crate as a library gets real error types)

jamesbornholt commented on mountpoint-s3/src/prefix.rs:
```diff
@@ -0,0 +1,65 @@
+        }
+    }
+
+    pub fn is_valid(prefix: &str) -> bool {
+        prefix.is_empty() || prefix.ends_with('/')
+    }
```
If you do the other comment in `lookup`, this doesn't need to be pub any more.

jamesbornholt commented on mountpoint-s3/src/prefix.rs:
```diff
@@ -0,0 +1,65 @@
+}
+
+impl FromStr for Prefix {
+    type Err = anyhow::Error;
```
Same thing here about the error type

jamesbornholt commented on mountpoint-s3/src/prefix.rs:
```diff
@@ -0,0 +1,65 @@
+
+impl Display for Prefix {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_fmt(format_args!("{}", self.path))
```
```suggestion
        write!(f, "{}", self.path)
```

passaro replied:
ah, much simpler! :)

jamesbornholt reviewed (commented):
Neat. I like this approach a bunch!

jamesbornholt commented on mountpoint-s3/src/prefix.rs:
```diff
@@ -0,0 +1,71 @@
+
+    use super::*;
+
+    #[test_case(" "; "whitespace")]
+    #[test_case("hello"; "not ending in slash")]
+    #[test_case("hello/world"; "nested folder not ending in slash")]
+    fn test_invalid_prefix(prefix: &str) {
+        assert!(!Prefix::is_valid(prefix))
+    }
+
+    #[test_case(""; "empty string")]
+    #[test_case("hello/"; "ending in slash")]
+    #[test_case("hello/world/"; "nested folder ending in slash")]
+    #[test_case(" /"; "whitespace ending in slash")]
+    #[test_case("/"; "single slash")]
+    #[test_case("//"; "double slash")]
+    #[test_case("/hello/"; "starting with slash")]
+    fn test_valid_prefix(prefix: &str) {
+        assert!(Prefix::is_valid(prefix))
+    }
```
One nitpick: for these tests we should test `Prefix::new` instead of `Prefix::is_valid`, which is now an implementation detail of `new`.

jamesbornholt commented on mountpoint-s3/src/prefix.rs:
```diff
@@ -0,0 +1,71 @@
+
+#[derive(Error, Debug)]
+pub enum PrefixError {
+    #[error("prefixes must end in '/'")]
```
Fixing my own grammar mistake: this should be "prefix must end in '/'"

jamesbornholt reviewed (commented):
Two more tiny things I missed, sorry!
